### PR TITLE
Remove DICT_KEYS guard

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118630
* #118855
* #118535
* #118199
* #118208
* #118003
* #118194
* #117625
* #117983
* #118098
* #117982

This guard was trying to do too much. We replace it for the weaker
DICT_CONST_KEYS, which is only valid when all the values are constants
(i.e. easy to print out).

We also implement an optimisation where, if all the keys to a dictionary
are constant, we generate just one guard of the form
```
list(d.keys()) == ["key1", "key2"]
```

This is the type of guards we used to generate initially before this set
of PRs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by lezcano** [comment]
The trickiest use of the DICT_KEYS guard was within `optimizer.py`.
There, this guard was used to "manually trace" through `optim.state`.
I have replaced that manual tracing with a call to `VariableBuilder`,
but that currently sets some dupe guards between `optim.state` and
`optim.param_groups["params"]`.

@mlazos do you originally wrote the code
this way to avoid that guard? If so, I can remove it and change it by a
`DICT_ID_KEYS` and still remove `DICT_KEYS`

**Review by peterbell10 — CHANGES_REQUESTED** [review]
Makes sense, but `test_compiled_optimizers.py` failure looks real.

**Comment by lezcano** [comment]
This fixed a graphbreak in asgd and nadam apparently. I have not looked into what happened there really, but I guess that's good.

File: torch/_dynamo/guards.py [hunk 1]
```diff
@@ -556,15 +532,15 @@ def NN_MODULE_PARAM_NAMES(self, guard):
 
         self._produce_guard_code(guard, code)
 
-    def ODICT_KEYS(self, guard):
-        """OrderedDict keys match"""
+    def DICT_CONST_KEYS(self, guard):
```
[the remark this answers is not in the document]
Comment (lezcano) [reply torch/_dynamo/guards.py:535]: I mean, there is a streak of 3 that are in alphabetical order... jajajajaj I'll leave it as is to keep the diff minimal.

**Comment by lezcano** [comment]
Triggered a benchmark run to see if there's any perf hit in all the changes in this stack.

**Review by jansel — APPROVED** [review]

**Review by mlazos — APPROVED** [review]

File: test/inductor/test_compiled_optimizers.py [hunk 2]
```diff
@@ -47,12 +47,16 @@ class KernelCounts(NamedTuple):
...
+    "test_nadam_foreach_cpu": 12,
+    "test_nadam_foreach_momentum_decay_cpu": 12,
+    "test_nadam_foreach_weight_decay_momentum_decay_decoupled_weight_decay_cpu": 12,
     "test_adamw_foreach_amsgrad_capturable_cuda": 3,
     "test_adamw_amsgrad_capturable_cuda": 6,
     "test_adam_amsgrad_capturable_cuda": 6,
     "test_adadelta_foreach_weight_decay_maximize_cpu": 12,
     "test_adadelta_foreach_rho_weight_decay_cpu": 12,
     "test_adadelta_foreach_weight_decay_cpu": 12,
+    "test_asgd_recompile_single": 16,
```
Comment (peterbell10) [inline test/inductor/test_compiled_optimizers.py:59]: NIT: I think this value is defined here

https://github.com/pytorch/pytorch/blob/44632939b5aa0c3e658d807c294df8fb3018fb91/test/inductor/test_compiled_optimizers.py#L320-L323

**Review by peterbell10 — APPROVED** [review]

**Comment by lezcano** [comment]
The optim changes seem to regress a bit. I'll look into how to get the graphbreak wins from this PR without regressing. I'm merging the rest of the stack, as that seems fine (might need to split the constant dict optimisation into its own PR tho as there are some regressions on gnns otherwise)

**Pull request closed without merging**